// @project-serum/anchor/dist/cjs/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.utils = exports.workspace = exports.AccountsCoder = exports.StateCoder = exports.EventCoder = exports.InstructionCoder = exports.Coder = exports.Wallet = exports.setProvider = exports.getProvider = exports.Provider = exports.web3 = exports.BN = void 0;
var bn_js_1 = require("bn.js");
Object.defineProperty(exports, "BN", { enumerable: true, get: function () { return __importDefault(bn_js_1).default; } });
exports.web3 = __importStar(require("@solana/web3.js"));
var provider_1 = require("./provider");
Object.defineProperty(exports, "Provider", { enumerable: true, get: function () { return __importDefault(provider_1).default; } });
Object.defineProperty(exports, "getProvider", { enumerable: true, get: function () { return provider_1.getProvider; } });
Object.defineProperty(exports, "setProvider", { enumerable: true, get: function () { return provider_1.setProvider; } });
Object.defineProperty(exports, "Wallet", { enumerable: true, get: function () { return provider_1.NodeWallet; } });
var coder_1 = require("./coder");
Object.defineProperty(exports, "Coder", { enumerable: true, get: function () { return __importDefault(coder_1).default; } });
Object.defineProperty(exports, "InstructionCoder", { enumerable: true, get: function () { return coder_1.InstructionCoder; } });
Object.defineProperty(exports, "EventCoder", { enumerable: true, get: function () { return coder_1.EventCoder; } });
Object.defineProperty(exports, "StateCoder", { enumerable: true, get: function () { return coder_1.StateCoder; } });
Object.defineProperty(exports, "AccountsCoder", { enumerable: true, get: function () { return coder_1.AccountsCoder; } });
__exportStar(require("./error"), exports);
var workspace_1 = require("./workspace");
Object.defineProperty(exports, "workspace", { enumerable: true, get: function () { return __importDefault(workspace_1).default; } });
exports.utils = __importStar(require("./utils"));
__exportStar(require("./program"), exports);
